DESIGN.md

```diff
@@ -0,0 +1,400 @@
+  - In case of remote verification implementation, the same model of [http.Request](https://pkg.go.dev/net/http#Request.WithContext) can be applied.
+- Primitive signature service providers can have their own implementation implementing `Signer` and / or `Verifier`.
+  - Remote signing a.k.a. KMS is one kind of primitive signature service providers.
+- `Headers` now stores raw CBOR object to avoid re-serialization in verification.
+  - The raw field enables advanced developers to marshal / unmarshal in their own way.
+- `I2OSP` now returns error. `OS2IP` is added as a pair of `I2OSP`.
+- `SignMessage.Verify` is kept but need to be updated as its verification policy is unclear.
+- More algorithm are supported.
+- `IsSign1Message` and `IsSignMessage` are removed since the motivation is unclear and cannot identify untagged messages.
+- `SignMessageCBORTag` and `Sign1MessageCBORTag` are renamed to `CBORTagSignMessage` and `CBORTagSign1Message` accordingly.
+
+### Examples
+
+Here are the example usage after the API changes are done.
+
+#### COSE_Sign1
+
+Sign a message:
+
+```go
+// create message to be signed
+msg := cose.NewSign1Message()
+msg.Payload = []byte("hello world")
+msg.Headers.Protected.SetAlgorithm(cose.AlgorithmES256)
+msg.Headers.Unprotected[cose.HeaderLabelKeyID] = 1
+
+// more efficient alternative
+msg = &cose.Sign1Message{
+    Headers: cose.Headers{
+        Protected: map[interface{}]interface{}{
+            cose.HeaderLabelAlgorithm: cose.AlgorithmES256.Value,
+        },
+        Unprotected: map[interface{}]interface{}{
+            cose.HeaderLabelKeyID: 1,
+        },
+    },
+    Payload: []byte("hello world"),
+}
+
+// sign message
+signer, _, err := cose.NewSignerWithEphemeralKey(cose.AlgorithmES256)
+check(err)
+err = msg.Sign(rand.Reader, nil, signer)
+check(err)
+sig, err := msg.MarshalCBOR()
+check(err)
+
+// remote signing scenario 1
+ctx := context.Background()
+signer = getRemoteSigner()
+signer = signer.WithContext(ctx)
+err = msg.Sign(rand.Reader, nil, signer)
+check(err)
+sig, err = msg.MarshalCBOR()
+check(err)
+
+// remote signing scenario 2
+signer = getRemoteSignerWithContext(ctx)
+err = msg.Sign(rand.Reader, nil, signer)
+check(err)
+sig, err = msg.MarshalCBOR()
+check(err)
+```
+
+Verify a message:
+
+```go
+rawSig := getRawSignature()
+key := getPublicKey()
+
+verifier, err := cose.NewVerifier(cose.AlgorithmES256, key)
+check(err)
+sig, err := cose.ParseSign1Message(rawSig)
+check(err)
+err = msg.Verify(nil, verifier)
+check(err) 
+```
+
+#### COSE_Sign
+
+Sign a message:
+
+```go
+// create signature holder
+sig := cose.NewSignature()
+sig.Headers.Protected.SetAlgorithm(cose.AlgorithmES256)
+sig.Headers.Unprotected[cose.HeaderLabelKeyID] = 1
+
+// more efficient alternative
+sig = &cose.Signature{
+    Headers: cose.Headers{
+        Protected: map[interface{}]interface{}{
+            cose.HeaderLabelAlgorithm: cose.AlgorithmES256.Value,
+        },
+        Unprotected: map[interface{}]interface{}{

```

Comment:
`Unprotected` is typed as `UnprotectedHeader`, so it should be instantiated as `Protected: cose.UnprotectedHeader{...}`.

Reply:
Good catch.